tensorflow/core/kernels/eigen_backward_spatial_convolutions.h

```diff
@@ -438,154 +400,100 @@ SpatialConvolutionBackwardKernel(
-
-  // Compute paddings for output_backward before extracting patches.
-  const TensorIndex expanded_out_rows = (outputRows - 1) * row_stride + 1;
-  const TensorIndex expanded_out_cols = (outputCols - 1) * col_stride + 1;
-
-  const TensorIndex padded_out_rows = inputRows + kernelRowsEff - 1;
-  const TensorIndex padded_out_cols = inputCols + kernelColsEff - 1;
-
-  const TensorIndex top_pad_rows = kernelRowsEff - 1 - padding_top;
-  const TensorIndex left_pad_cols = kernelColsEff - 1 - padding_left;
-
-  const TensorIndex bottom_pad_rows =
-      padded_out_rows - expanded_out_rows - top_pad_rows;
-  const TensorIndex right_pad_cols =
-      padded_out_cols - expanded_out_cols - left_pad_cols;
-
-  // Reorder output_backward dimensions.
-  array<TensorIndex, 4> output_backward_shuffle;
-  if (isColMajor) {
-    // From: [out_depth, out_rows, out_cols, batch]
-    // To:   [batch, out_rows, out_cols, out_depth]
-    output_backward_shuffle = {3, 1, 2, 0};
-  } else {
-    // From: [batch, out_cols, out_rows, out_depth]
-    // To:   [out_depth, out_cols, out_rows, batch]
-    output_backward_shuffle = {3, 1, 2, 0};
-  }
-
-  // Reorder input dimensions.
-  array<TensorIndex, 4> input_shuffle;
-  if (isColMajor) {
-    // From: [in_depth, in_rows, in_cols, batch]
-    // To:   [in_depth, batch, in_rows, in_cols]
-    input_shuffle = {0, 3, 1, 2};
-  } else {
-    // From: [batch, in_cols, in_rows, in_depth]
-    // To:   [in_cols, in_rows, batch, in_depth]
-    input_shuffle = {1, 2, 0, 3};
-  }
-
-  // Input is playing the role of a "kernel" in this convolution.
-  DSizes<TensorIndex, 2> input_dims;
+  // Reshaped out
+  DSizes<TensorIndex, 2> output_dims;
   if (isColMajor) {
-    input_dims[0] = kernelChannels;
-    input_dims[1] = batch * inputRows * inputCols;
+    output_dims[0] = kernelFilters;
+    output_dims[1] = outputRows * outputCols;
+    for (int i = 3; i < NumDims; ++i) {
+      output_dims[1] *= out.dimension(i);
+    }
   } else {
-    input_dims[1] = kernelChannels;
-    input_dims[0] = inputCols * inputRows * batch;
+    output_dims[1] = kernelFilters;
+    output_dims[0] = outputCols * outputRows;
+    for (int i = 0; i < NumDims - 3; ++i) {
+      output_dims[0] *= out.dimension(i);
+    }
   }
 
-  // Molds the output of the patch extraction result into a 2D tensor:
-  // - the first dimension (dims[0]): the patch values to be multiplied with the
-  // kernels
-  // - the second dimension (dims[1]): everything else
+  // Reshaped extract_image_patches(in)
   DSizes<TensorIndex, 2> pre_contract_dims;
   if (isColMajor) {
-    pre_contract_dims[0] = batch * inputRows * inputCols;
-    pre_contract_dims[1] = kernelRows * kernelCols * kernelFilters;
+    pre_contract_dims[0] = kernelChannels * kernelRows * kernelCols;
+    pre_contract_dims[1] = outputRows * outputCols;
+    for (int i = 3; i < NumDims; ++i) {
+      pre_contract_dims[1] *= in.dimension(i);
+    }
+    eigen_assert(output_dims[1] == pre_contract_dims[1]);
```

Comment:
I don't think we need the asserts here.  I believe if there was a mismatch (and if asserts were enabled) it would be caught in the actual returned expression.

Reply:
This part was completely reused from TF 1.13, as it was too different from the current code, that might explain some discrepancies in the style.